# <center> Segmenting and Clustering Toronto Neighbourhoods <center>

# Objective: Segment and cluster the Toronto neighbourhoods based on post codes

Download the table of post codes for neighbourhoods in Toronto from Wikipedia. https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M 

I have Excel 2016, which can hold 1,048,576 rows. The post code table has only 287 rows. I tried using BeautifulSoup first. But when I was half-way through and dealing with all those issues of cleaning the data, I realized that for such a small table, directlying downloading it into Excel, then uploading it into Jupyter Notebook will be much easier and faster as well. So, I scraped all my BeautifulSoup codes.

In [127]:
import pandas as pd
postcode_df=pd.read_excel('Toronto Post Codes.xlsx')
postcode_df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [128]:
# size of the table
postcode_df.shape

(287, 3)

In [129]:
# number of unique Boroughs
import numpy as np
postcode_df["Borough"].value_counts()

Not assigned        77
Etobicoke           45
North York          38
Scarborough         37
Downtown Toronto    37
Central Toronto     17
West Toronto        13
York                 9
East Toronto         7
East York            6
Mississauga          1
Name: Borough, dtype: int64

Okay. So there are 77 "Not assigned" in the Borough column. I could use drop.duplicates() to drop those "Not assigned". However, that will drop the other boroughs as well, since there are boroughs that have more than one neighbourhoods assigned to it. So, drop by "Borough" won't work.

Let's check the Neighbourhood column. The neighbourhoods should be unique, except for the "Not assigned".

In [130]:
postcode_df["Neighbourhood"].value_counts()

Not assigned                                         77
St. James Town                                        2
Runnymede                                             2
Mount Dennis                                          1
York Mills West                                       1
The Queensway East                                    1
Royal York South East                                 1
Martin Grove                                          1
Long Branch                                           1
Wexford                                               1
East Birchmount Park                                  1
Adelaide                                              1
Woodbine Gardens                                      1
The Queensway West                                    1
Humber Summit                                         1
Woburn                                                1
Clarks Corners                                        1
Cloverdale                                      

All the neighbourhoods are unique, except for 1) the "Not assigned" and  2) "Runnymede" and "St. James Town" where there are two copies.

In [131]:
#Remove the "Not assigned" from the Borough column.
postcode_df.drop_duplicates(subset="Neighbourhood",keep=False, inplace=True)
postcode_df

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
9        M9A         Etobicoke   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
149      M9N              York   
150      M1P       Scarborough   
151      M1P       Scarborough   
152      M1P       Scarbor

All the "Not assigned" in the Neighbourhood column has been dropped. However, since the parameter "keep" was set to False. The process above dropped the two neighbourhoods "Runnemede" and "St. James Town" as well. But I would like to keep those two neighbourhoods, since each copy belong to two different boroughs.

Put the two neighbourhoods back.

In [133]:
append1=pd.DataFrame({"Postcode":["M5C","M6N","M6S","M4X"],"Borough":["Downtown Toronto","York","West Toronto","Downtown Toronto"],"Neighbourhood":["St. James Town","Runnymede","Runnymede","St. James Town"]})
postcode_df.append(append1,ignore_index=False)

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
9        M9A         Etobicoke   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
149      M9N              York   
150      M1P       Scarborough   
151      M1P       Scarborough   
152      M1P       Scarbor

In [138]:
print("The cleaned table has", len(postcode_df.append(append1, ignore_index=False)), "row")

The cleaned table has 210 row


As shown above,the original table has 287 rows in total and 77 "Not assigned" in the Borough column. After the 77 "Not assigned" rows are dropped, the cleaned table should have 287-77=210 rows. The result of the above code block shows 210 rows. 

Let's confirm that there is no more "Not assigned" in Borough column.

In [139]:
postcode_df["Borough"].value_counts()

Etobicoke           45
North York          38
Scarborough         37
Downtown Toronto    35
Central Toronto     17
West Toronto        12
York                 8
East Toronto         7
East York            6
Mississauga          1
Name: Borough, dtype: int64

Cool. There is no more "Not assigned" in the Borough column. 

The table is cleaned now. I can start working with it. The next thing I am going to do is to put the neighbourhoods that have the same postcode in the same row. Let's see how many unique postcodes are in the cleaned table.

In [152]:
count=pd.DataFrame(postcode_df["Postcode"].value_counts())
print("There are ",len(count)," unique post codes.")

There are  102  unique post codes.


There are 102 unique postcodes. Most of the postcodes has only 1 neighbourhoods associated with it. However, postcodeS "M8Y" and "M9V" have 8 neighbourhoods associated with it, "M5V" has 7, "M4V" and "M8Z" have 5. I need to concatenate all those neighbourhoods under the postcode they are associated with, in one row.

In [153]:
postcode_df=pd.DataFrame(postcode_df.groupby(["Postcode","Borough"])["Neighbourhood"].apply(lambda x: ','.join(x)))
postcode_df

Neighbourhood
Postcode Borough                                                            
M1B      Scarborough                                           Rouge,Malvern
M1C      Scarborough                    Highland Creek,Rouge Hill,Port Union
M1E      Scarborough                         Guildwood,Morningside,West Hill
M1G      Scarborough                                                  Woburn
M1H      Scarborough                                               Cedarbrae
M1J      Scarborough                                     Scarborough Village
M1K      Scarborough               East Birchmount Park,Ionview,Kennedy Park
M1L      Scarborough                           Clairlea,Golden Mile,Oakridge
M1M      Scarborough           Cliffcrest,Cliffside,Scarborough Village West
M1N      Scarborough                              Birch Cliff,Cliffside West
M1P      Scarborough       Dorset Park,Scarborough Town Centre,Wexford He...
M1R      Scarborough                                        Maryvale,Wexford
M1S      Scarborough                                               Agincourt
M1T      Scarborough                   Clarks Corners,Sullivan,Tam O'Shanter
M1V      Scarborough       Agincourt North,L'Amoreaux East,Milliken,Steel...
M1W      Scarborough                                         L'Amoreaux West
M1X      Scarborough                                             Upper Rouge
M2H      North York                                        Hillcrest Village
M2J      North York                               Fairview,Henry Farm,Oriole
M2K      North York                                          Bayview Village
M2L      North York                                  Silver Hills,York Mills
M2M      North York                                   Newtonbrook,Willowdale
M2N      North York                                         Willowdale South
M2P      North York                                          York Mills West
M2R      North York                                          Willowdale West
M3A      North York                                                Parkwoods
M3B      North York                                          Don Mills North
M3C      North York                          Flemingdon Park,Don Mills South
M3H      North York            Bathurst Manor,Downsview North,Wilson Heights
M3J      North York                           Northwood Park,York University
M3K      North York                               CFB Toronto,Downsview East
M3L      North York                                           Downsview West
M3M      North York                                        Downsview Central
M3N      North York                                      Downsview Northwest
M4A      North York                                         Victoria Village
M4B      East York                            Woodbine Gardens,Parkview Hill
M4C      East York                                          Woodbine Heights
M4E      East Toronto                                            The Beaches
M4G      East York                                                   Leaside
M4H      East York                                          Thorncliffe Park
M4J      East York                                              East Toronto
M4K      East Toronto                            The Danforth West,Riverdale
M4L      East Toronto                          The Beaches West,India Bazaar
M4M      East Toronto                                        Studio District
M4N      Central Toronto                                       Lawrence Park
M4P      Central Toronto                                    Davisville North
M4R      Central Toronto                                  North Toronto West
M4S      Central Toronto                                          Davisville
M4T      Central Toronto                          Moore Park,Summerhill East
M4V      Central Toronto   Deer Park,Forest Hill SE,Rathnelly,South Hill,...
M4W      Downtown Toronto                                  

All the neighbourhoods are now concatenated under the postcode they are associated with. However, instead of regular index, the table has "Postcode" and "Borough" as its multi-column index. I am going to fix it below.

In [154]:
postcode_df.reset_index(inplace=True)
postcode_df

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5E  Downtown Toronto   
56       M5G  Downtown Toronto   
57       M5H  Downtown Toronto   
58       M5J  Downtown Toronto   
59       M5K  Downtown Toronto   
60       M5L  Downtown Toronto   
61       M5M        North York   
62       M5N   Central Toronto   
63       M5P   Central Toronto   
64       M5R   Central Toronto   
65       M5S  Downtown Toronto   
66       M5T  Downtown Toronto   
67       M5V  Downtown Toronto   
68       M5W  Downtown Toronto   
69       M5X  Downtown Toronto   
70       M6A        North York   
71       M6B        North York   
72       M6C              York   
73       M6E              York   
74       M6G  Downtown Toronto   
75       M6H      West Toronto   
76       M6J      West Toronto   
77       M6K      West Toronto   
78       M6L        North York   
79       M6M              York   
80       M6N              York   
81       M6P      West Toronto   
82       M6R      West Toronto   
83       M6S      West Toronto   
84       M7A  Downtown Toronto   
85       M7R       Mississauga   
86       M7Y      East Toronto   
87       M8V         Etobicoke   
88       M8W         Etobicoke   
89       M8X         Etobicoke   
90       M8Y         Etobicoke   
91       M8Z         Etobicoke   
92       M9A         Etobicoke   
93       M9B         Etobicoke   
94       M9C         Etobicoke   
95       M9L        North York   
96       M9M        North York   
97       M9N              York   
98       M9P         Etobicoke   
99       M9R         Etobicoke   
100      M9V         Etobicoke   
101      M9W         Etobicoke   

                                         Neighbourhood  
0                                        Rouge,Malvern  
1                 Highland Creek,Rouge Hill,Port Union  
2                      Guildwood,Morningside,West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6            East Birchmount Park,Ionview,Kennedy Park  
7                        Clairlea,Golden Mile

In [155]:
print(postcode_df.shape)

(102, 3)


Next, find the latitude and longitude coordinates for each postcode.

In [107]:
# Load the "Geospatial_Coordinates.csv" dataset
Geo_df=pd.read_csv("Geospatial_Coordinates.csv")
Geo_df.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

In [108]:
# Merge "postcode_df" and "Geo_df". 
# Use "left" join so that all records from "postcode_df" and matched records from "Geo_df" are returned.
Geomerge_df=pd.merge(postcode_df,Geo_df, how="left")
Geomerge_df

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5E  Downtown Toronto   
56       M5G  Downtown Toronto   
57       M5H  Downtown Toronto   
58       M5J  Downtown Toronto   
59       M5K  Downtown Toronto   
60       M5L  Downtown Toronto   
61       M5M        North York   
62       M5N   Central Toronto   
63       M5P   Central Toronto   
64       M5R   Central Toronto   
65       M5S  Downtown Toronto   
66       M5T  Downtown Toronto   
67       M5V  Downtown Toronto   
68       M5W  Downtown Toronto   
69       M5X  Downtown Toronto   
70       M6A        North York   
71       M6B        North York   
72       M6C              York   
73       M6E              York   
74       M6G  Downtown Toronto   
75       M6H      West Toronto   
76       M6J      West Toronto   
77       M6K      West Toronto   
78       M6L        North York   
79       M6M              York   
80       M6N              York   
81       M6P      West Toronto   
82       M6R      West Toronto   
83       M6S      West Toronto   
84       M7A  Downtown Toronto   
85       M7R       Mississauga   
86       M7Y      East Toronto   
87       M8V         Etobicoke   
88       M8W         Etobicoke   
89       M8X         Etobicoke   
90       M8Y         Etobicoke   
91       M8Z         Etobicoke   
92       M9A         Etobicoke   
93       M9B         Etobicoke   
94       M9C         Etobicoke   
95       M9L        North York   
96       M9M        North York   
97       M9N              York   
98       M9P         Etobicoke   
99       M9R         Etobicoke   
100      M9V         Etobicoke   
101      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                          

Cluster the neighbourhoods and create visualizations.

Import the necessary libraries first.

In [109]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Get the latitude and longitude for Toronto City.

In [111]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Canada are 43.653963, -79.387207.


Create a map of Toronto City with the neighbourhoods superimposed on it.

In [112]:
# Toronto latitude and longitude.
latitude= 43.653963
longitude=-79.387207

# create map and display it
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=12)

# instantiate a feature group for the incidents in the dataframe
incidents = folium.map.FeatureGroup()

# loop through the 100 crimes and add each to the incidents feature group
for lat, lng, in zip(Geomerge_df.Latitude, Geomerge_df.Longitude):
    incidents.add_child(
        folium.features.CircleMarker(
            [lat, lng],
            radius=5, # define how big you want the circle markers to be
            color='red',
            fill=True,
            fill_color='blue',
            fill_opacity=0.6
        )
    )

# add incidents to map
toronto_map.add_child(incidents)

Simplify the above map and segment and cluster only the neighborhoods with "Toronto" in their borough names. These are the neighbourhoods associated with the following boroughs: Central Toronto, Downtown Toronto, East Toronto and West Toronta. So let's slice the original dataframe and create a new dataframe for the Toronto boroughs.

In [113]:
toronto_data1 = Geomerge_df[Geomerge_df['Borough'] == 'Central Toronto'].reset_index(drop=True)
toronto_data2 = Geomerge_df[Geomerge_df['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
toronto_data3 = Geomerge_df[Geomerge_df['Borough'] == 'East Toronto'].reset_index(drop=True)
toronto_data4 = Geomerge_df[Geomerge_df['Borough'] == 'West Toronto'].reset_index(drop=True)
frames=[toronto_data1,toronto_data2,toronto_data3,toronto_data4]
toronto_data=pd.concat(frames)     
toronto_data.head()

Postcode          Borough               Neighbourhood   Latitude  Longitude
0      M4N  Central Toronto               Lawrence Park  43.728020 -79.388790
1      M4P  Central Toronto            Davisville North  43.712751 -79.390197
2      M4R  Central Toronto          North Toronto West  43.715383 -79.405678
3      M4S  Central Toronto                  Davisville  43.704324 -79.388790
4      M4T  Central Toronto  Moore Park,Summerhill East  43.689574 -79.383160

Let's get the latitude and longitude for Downtown Toronto.

Visualize the Toronto boroughs with their neighbourhoods.

In [114]:

# create map of Manhattan using latitude and longitude values
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)  
    
toronto_map

Let's cluster the neighbourhoods in Downtown Toronto.

In [115]:
# Onehot coding
toronto_data_onehot = pd.get_dummies(toronto_data[['Neighbourhood']], prefix="", prefix_sep="")

# add Borough column back to dataframe
toronto_data_onehot['Borough'] = toronto_data['Borough'] 

# move neighborhood column to the first column
fixed_columns = [toronto_data_onehot.columns[-1]] + list(toronto_data_onehot.columns[:-1])
toronto_data_onehot = toronto_data_onehot[fixed_columns]

toronto_data_onehot.head(10)

Borough  Adelaide,King,Richmond  Berczy Park  \
0   Central Toronto                       0            0   
1   Central Toronto                       0            0   
2   Central Toronto                       0            0   
3   Central Toronto                       0            0   
4   Central Toronto                       0            0   
5   Central Toronto                       0            0   
6   Central Toronto                       0            0   
7   Central Toronto                       0            0   
8   Central Toronto                       0            0   
0  Downtown Toronto                       0            0   

   Brockton,Exhibition Place,Parkdale Village  \
0                                           0   
1                                           0   
2                                           0   
3                                           0   
4                                           0   
5                                           0   
6                                           0   
7                                           0   
8                                           0   
0                                           0   

   Business Reply Mail Processing Centre 969 Eastern  \
0                                                  0   
1                                                  0   
2                                                  0   
3                                                  0   
4                                                  0   
5                                                  0   
6                                                  0   
7                                                  0   
8                                                  0   
0                                                  0   

   CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara  \
0                                                  0                                                      
1                                                  0                                                      
2                                                  0                                                      
3                                                  0                                                      
4                                                  0                                                      
5                                                  0                                                      
6                                                  0                                                      
7                                                  0                                                      
8                                                  0                                                      
0                                                  0                                                      

   Cabbagetown  Central Bay Street  Chinatown,Grange Park,Kensington Market  \
0            0                   0                                        0   
1            0                   0                                        0   
2            0                   0                                        0   
3            0                   0                                        0   
4            0                   0                                        0   
5            0                   0                                        0   
6            0                   0                                        0   
7            0                   0                                        0   
8            0                   0                                        0   
0            0                   0                                        0   

   Christie  Church and Wellesley  Commerce Court,Victoria Hotel  Davisville  \
0         0                     0                              0  

Group the rows by boroughs and by taking the mean of the frequency of occurrence of each category.

In [116]:
toronto_grouped =toronto_data_onehot.groupby('Borough').mean().reset_index()
toronto_grouped

Borough  Adelaide,King,Richmond  Berczy Park  \
0   Central Toronto                0.000000     0.000000   
1  Downtown Toronto                0.055556     0.055556   
2      East Toronto                0.000000     0.000000   
3      West Toronto                0.000000     0.000000   

   Brockton,Exhibition Place,Parkdale Village  \
0                                    0.000000   
1                                    0.000000   
2                                    0.000000   
3                                    0.166667   

   Business Reply Mail Processing Centre 969 Eastern  \
0                                                0.0   
1                                                0.0   
2                                                0.2   
3                                                0.0   

   CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara  \
0                                           0.000000                                                      
1                                           0.055556                                                      
2                                           0.000000                                                      
3                                           0.000000                                                      

   Cabbagetown  Central Bay Street  Chinatown,Grange Park,Kensington Market  \
0     0.000000            0.000000                                 0.000000   
1     0.055556            0.055556                                 0.055556   
2     0.000000            0.000000                                 0.000000   
3     0.000000            0.000000                                 0.000000   

   Christie  Church and Wellesley  Commerce Court,Victoria Hotel  Davisville  \
0  0.000000              0.000000                       0.000000    0.111111   
1  0.055556              0.055556                       0.055556    0.000000   
2  0.000000              0.000000                       0.000000    0.000000   
3  0.000000              0.000000                       0.000000    0.000000   

   Davisville North  \
0          0.111111   
1          0.000000   
2          0.000000   
3          0.000000   

   Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West  \
0                                           0.111111               
1                                           0.000000               
2                                           0.000000               
3                                           0.000000               

   Design Exchange,Toronto Dominion Centre  Dovercourt Village,Dufferin  \
0                                 0.000000                     0.000000   
1                                 0.055556                     0.000000   
2                                 0.000000                     0.000000   
3                                 0.000000                     0.166667   

   First Canadian Place,Underground city  Forest Hill North,Forest Hill West  \
0                               0.000000                            0.111111   
1                               0.055556                            0.000000   
2                               0.000000                            0.000000   
3                               0.000000                            0.000000   

   Harbord,University of Toronto  Harbourfront  \
0                       0.000000      0.000000   
1                       0.055556      0.055556   
2                       0.000000      0.000000   
3                       0.000000      0.000000   

   Harbourfront East,Toronto Islands,Union Station  \
0                                         0.000000   
1                                         0.055556   
2                                         0.000000   
3                                         0.000000   

   High Park,The Junction South  Lawrence Park  Little Portugal,Trinity  \
0                   

Cluster the neighbourhoods in Central Toronto, Downtown Toronto, East Toronto and West Toronto.

In [117]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 4

toronto_clustering = toronto_grouped.drop('Borough', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:3] 

array([0, 3, 1])

Visualize the clusters of the neighbourhoods in Central Toronto, Downtown Toronto, East Toronto and West Toronto on the map.

In [118]:
# Add cluster labels
toronto_grouped.insert(0,"Cluster label",kmeans.labels_)

In [119]:
toronto_data=toronto_data.join(toronto_grouped.set_index("Borough"),on="Borough")
toronto_data.head()

Postcode          Borough               Neighbourhood   Latitude  Longitude  \
0      M4N  Central Toronto               Lawrence Park  43.728020 -79.388790   
1      M4P  Central Toronto            Davisville North  43.712751 -79.390197   
2      M4R  Central Toronto          North Toronto West  43.715383 -79.405678   
3      M4S  Central Toronto                  Davisville  43.704324 -79.388790   
4      M4T  Central Toronto  Moore Park,Summerhill East  43.689574 -79.383160   

   Cluster label  Adelaide,King,Richmond  Berczy Park  \
0              0                     0.0          0.0   
1              0                     0.0          0.0   
2              0                     0.0          0.0   
3              0                     0.0          0.0   
4              0                     0.0          0.0   

   Brockton,Exhibition Place,Parkdale Village  \
0                                         0.0   
1                                         0.0   
2                                         0.0   
3                                         0.0   
4                                         0.0   

   Business Reply Mail Processing Centre 969 Eastern  \
0                                                0.0   
1                                                0.0   
2                                                0.0   
3                                                0.0   
4                                                0.0   

   CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara  \
0                                                0.0                                                      
1                                                0.0                                                      
2                                                0.0                                                      
3                                                0.0                                                      
4                                                0.0                                                      

   Cabbagetown  Central Bay Street  Chinatown,Grange Park,Kensington Market  \
0          0.0                 0.0                                      0.0   
1          0.0                 0.0                                      0.0   
2          0.0                 0.0                                      0.0   
3          0.0                 0.0                                      0.0   
4          0.0                 0.0                                      0.0   

   Christie  Church and Wellesley  Commerce Court,Victoria Hotel  Davisville  \
0       0.0                   0.0                            0.0    0.111111   
1       0.0                   0.0                            0.0    0.111111   
2       0.0                   0.0                            0.0    0.111111   
3       0.0                   0.0                            0.0    0.111111   
4       0.0                   0.0                            0.0    0.111111   

   Davisville North  \
0          0.111111   
1          0.111111   
2          0.111111   
3          0.111111   
4          0.111111   

   Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West  \
0                                           0.111111               
1                                           0.111111               
2                                           0.111111               
3                                           0.111111               
4                                           0.111111               

   Design Exchange,Toronto Dominion Centre  Dovercourt Village,Dufferin  \
0                                      0.0                          0.0   
1                                      0.0                          0.0   
2                                      0.0                          0.0   
3                                      0.0                          0.0   
4                 

In [120]:

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11.5)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'],toronto_data['Cluster label']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.8).add_to(map_clusters)
       
map_clusters

There we have the map of Toronto City with the four boroughs with Toronton in their names (Central Toronto, Downtown Toronto, East Toronto and West Toronto) superimposed on it. Each color represents a different borough, while each dot represents a neighbbourhood. To see which borough and cluster a dot represents, hover the cursor over the dot and click on it.

# Thanks for reading!